TFNetworkLayer.py

```diff
@@ -1892,14 +1893,19 @@ def __init__(self, activation, with_bias=True, grad_filter=None,
     :param str forward_weights_init: see :func:`TFUtil.get_initializer`
     :param str recurrent_weights_init: see :func:`TFUtil.get_initializer`
     :param str|float bias_init: see :func:`TFUtil.get_initializer`
+    :param int|None out_feature_dim: if set, will convert feature dim of the input and output data
     """
     super(LinearLayer, self).__init__(**kwargs)
     from TFUtil import get_initializer
 
     self.activation = activation
     self.with_bias = with_bias
 
-    input_data = self.input_data
+    if out_feature_dim is None:
+      input_data = self.input_data
+    else:
+      input_data = self.input_data.copy_with_feature_dim_axis(out_feature_dim)
+      self.output = self.output.copy_with_feature_dim_axis(out_feature_dim)
```

Comment:
The format of `self.output` should not be changed here. This should be done via overwriting `get_out_data_from_opts`.

Reply:
Actually, maybe better would be to extend the logic in `LayerBase._base_get_out_data_from_opts`.

Reply:
I mean for the case that the input data is channel first.

Reply:
Ok

Reply:
In the base class it's to abstract to decide what the output should look like. So, just coping from the input is ok. 
I'l do the output transformation in get_out_data_from_opts.

Reply:
In the base class, it will just copy the input format by default (i.e. output will be batch major if input is batch major; or output will be time major if input is time major). That is exactly what you also want here. So why not just add this additional default behavior?

Reply:
It may be the case, that the input is in NCHW, but it should be transformed to the NHWC due to the two reasons:
- Channels first are not supported later in the loss function (https://github.com/rwth-i6/returnn/blob/1a1d901b75e7175a999d469891ccda24de9f10ae/TFNetworkLayer.py#L5093). And the labels have channels on the last position.
- In the LinearLayer there's a function dot, which assumes features in the last axis (https://github.com/rwth-i6/returnn/blob/1a1d901b75e7175a999d469891ccda24de9f10ae/TFUtil.py#L2168). I think, it's possible to implement dot(a, b, feature_axis). But it's not needed in my case. In the official ResNet implementation, they transform the data flow to NHWC right before the last linear layer and softmax.

Reply:
You should not add a workaround in some unrelated layer just because that would not be supported in the loss function. Then just fix the loss function instead.
About the `dot` function: Also see `DotLayer`, which is more generic and has implemented sth like that already.